### ISSUE TITLE
WGPU Backend

### ISSUE THREAD

**Issue by undersquire** [not-fl3/macroquad#342]
This isn't an urgent issue, however, Apple has deprecated OpenGL on all of their platforms, meaning that they could simply drop support for it at any time with any new release. This would break macroquad's support for these platforms.

So I am curious if there are any plans to add support for a WGPU backend, since it supports many various backends under the hood including OpenGL, DirectX, Metal, Vulkan, etc. This can allow the user to choose what backend to compile with in order to build for certain platforms, for example, using DirectX for windows, Metal for Apple's platforms, and OpenGL or Vulkan for everything else.

**Comment by alexmozaidze** [comment]
wgpu has a best-effort support for OpenGL, which means that some OpenGL graphics cards that lack certain functionality cannot be supported by wgpu (aka my integrated graphics). It would be nice to have wgpu as a default option, and the old renderer as a fallback.

**Comment by undersquire** [comment]
I agree, I think having WGPU as the default with an OpenGL fallback would be the best way to go.

**Comment by Edersteiner** [comment]
Metal support would be greatly appreciated. OpenGL just doesn't have a bright future (or any future at all) on MacOS. Apple Silicon also runs Metal apps like a dream.

**Comment by NHodgesVFX** [comment]
I think a WGPU back end would also be good as well. People here have already mentioned the apple issues with OpenGL, so I wanted to look at some other reasons. Having WGPU as a backend options means it would be possible to do more advanced things like compute shaders, even if it doesn't get exposed directly I assume you would still be able to access it. WGPU could also bring better performace as well.

One thing I did notice though was the WGPU seems to be larger in size where macroquad with its current renderer on web seems to be around 500kb WGPU is around 2mb. So it would be nice to have both renderers available so you can pick and choose whats best for your project. 

Granted having two renderers seems like it would be a lot to maintain though.

**Comment by lenscas** [comment]
> One thing I did notice though was the WGPU seems to be larger in size where macroquad with its current renderer on web seems to be around 500kb WGPU is around 2mb. So it would be nice to have both renderers available so you can pick and choose whats best for your project.

I haven't looked at this at all yet but... is the size difference really that big on WASM, especially on browsers that support webgpu? (I know, still early but oh well). Sounds like all it would need to do in that environment is acting as bindings to the webgpu API...

**Comment by miwarnec** [comment]
Bump: macroquad for wgpu would be super valuable.

**Comment by inevity** [comment]
why no any progress?

**Comment by lenscas** [comment]
> why no any progress?

Because no one felt compelled to work on it.
